cpp/src/arrow/python/numpy_convert.cc

```diff
@@ -274,5 +275,348 @@ Status TensorToNdarray(const std::shared_ptr<Tensor>& tensor, PyObject* base,
   return Status::OK();
 }
 
+Status SparseTensorCOOToNdarray(const std::shared_ptr<SparseTensorCOO>& sparse_tensor,
+                                PyObject* base, PyObject** out_data,
+                                PyObject** out_coords) {
+  PyAcquireGIL lock;
+
+  int type_num_data;
+  int type_num_coords;
+  const auto& sparse_index = arrow::internal::checked_cast<const SparseCOOIndex&>(
+      *sparse_tensor->sparse_index());
+  const std::shared_ptr<arrow::NumericTensor<arrow::Int64Type>> sparse_index_coords =
+      sparse_index.indices();
+
+  RETURN_NOT_OK(GetNumPyType(*sparse_tensor->type(), &type_num_data));
+  PyArray_Descr* dtype_data = PyArray_DescrNewFromType(type_num_data);
+  RETURN_NOT_OK(GetNumPyType(*sparse_index_coords->type(), &type_num_coords));
+  PyArray_Descr* dtype_coords = PyArray_DescrNewFromType(type_num_coords);
+  RETURN_IF_PYERROR();
+
+  const int ndim_coords = sparse_tensor->ndim();
+  std::vector<npy_intp> npy_shape_coords(ndim_coords);
+
+  for (int i = 0; i < ndim_coords; ++i) {
+    npy_shape_coords[i] = sparse_index_coords->shape()[i];
+  }
+
+  std::vector<npy_intp> npy_shape_data({sparse_index.non_zero_length(), 1});
+
+  const void* immutable_data = nullptr;
+  if (sparse_tensor->data()) {
+    immutable_data = sparse_tensor->data()->data();
+  }
+
+  const void* immutable_coords = nullptr;
+  if (sparse_index_coords->data()) {
+    immutable_coords = sparse_index_coords->data()->data();
+  }
+
+  // Remove const =(
+  void* mutable_data = const_cast<void*>(immutable_data);
+  void* mutable_coords = const_cast<void*>(immutable_coords);
+
+  int array_flags = 0;
+  if (sparse_tensor->is_mutable()) {
+    array_flags |= NPY_ARRAY_WRITEABLE;
+  }
+
+  PyObject* result_data =
+      PyArray_NewFromDescr(&PyArray_Type, dtype_data, 1, npy_shape_data.data(), nullptr,
+                           mutable_data, array_flags, nullptr);
+  PyObject* result_coords =
+      PyArray_NewFromDescr(&PyArray_Type, dtype_coords, 2, npy_shape_coords.data(),
```

Comment:
Looks wrong. You're passing `2` for ndims but `npy_shape_coords` has `ndim_coords` elements.